File: leetcode/388-longest-absolute-file-path.py
class Solution(object):
    def lengthLongestPath(self, input):
        """
        :type input: str
        :rtype: int
        """
        maxlen = 0
        stk = [0]
        curlev, isfile, cnt = 0, 0, 0
        for c in input+'\n':
            if '\n' == c:
                while len(stk) > curlev + 1:
                    stk.pop()
                stk.append(cnt + stk[-1])
                if isfile:
                    isfile = 0
                    if stk[-1] + curlev > maxlen:
                        maxlen = stk[-1] + curlev
                curlev = 0
                cnt = 0
            elif '\t' == c:
                curlev += 1
            else:
                cnt += 1
                if '.' == c:
                    isfile = 1
        return maxlen

File: leetcode/388-longest-absolute-file-path.js
/**
 * @param {string} input
 * @return {number}
 */
var lengthLongestPath = function(input) {
  var maxlen = 0;
  var text = input + '\n';
  var stk = [0];
  for (var i = 0, cnt = 0, curlev = 0, isfile = 0;
       i < text.length; ++i) {
    switch (text[i]) {
    case '\n':
      while (stk.length > curlev + 1)
        stk.pop();
      stk.push(cnt + stk[stk.length-1]);
      if (isfile) {
        isfile = 0;
        if (curlev + stk[stk.length-1] > maxlen)
          maxlen = curlev + stk[stk.length-1];
      }
      curlev = 0;
      cnt = 0;
      break;
    case '\t':
      ++curlev;
      break;
    case '.':
      isfile = 1;
    default:
      ++cnt;
      break;
    }
  }
  return maxlen;
};
